feature: added swap and calculateswap functions to sdk

CryptoMaxPlanck opened the pull request:
the goal of this PR is to give developers the ability to utilize Swap.sol's swap function via an SDK

CryptoMaxPlanck committed:
feature: added swap and calculateswap functions to sdk

CryptoMaxPlanck commented on src/hooks/amm.ts:
```diff
@@ -0,0 +1,93 @@
+  deadline?: number
+) {
+  // Get the swap contract
+  const { swapAddress, swapEthAddress } = usePoolTokenInfo(poolName, chainId);
```
currently `usePoolTokenInfo()` is not a correct way to get the swapAddress. waiting for the PR from metagabbar with the tokens and chains to add

bholagabbar replied:
Isn't `hooks` react specific ? I thought we wanted this SDK to cater to general JS bridging functionality.

CryptoMaxPlanck replied:
as in the dir that the file is in? i kind of just put it here cause I saw bridge.ts here too, but probably would be better if i moved this to its own folder and kept everything more organized in src

CryptoMaxPlanck replied:
resolving: i think this is the right place to keep it? i don't think hooks are only react specific, but might be wrong

CryptoMaxPlanck commented on src/hooks/index.ts:
```diff
@@ -1,3 +1,5 @@
+export * from "./amm";
+
 export {
```
should we use * for the rest of the export? or is it better to specify the functions/types we want to export

bholagabbar replied:
Explicit imports

CryptoMaxPlanck committed:
fix:
- add functions to index
- clean utils

CryptoMaxPlanck commented:
for getting the swap contract address per chain, i am thinking about changing the SynapseContract class defined in src/common/synapse_contracts.ts to add a `swap` field and storing each chain's `Swap.sol` address there. this will make it easier to find the contract address when using ethers to get a contract instance

CryptoMaxPlanck committed:
add:
- change synapse_contract to support swap contract
- TODO: add swap contract addresses
- change amm code to get contract address from synapse_contract address

CryptoMaxPlanck committed:
Merge branch 'feature/amm' into feature/swap

CryptoMaxPlanck commented on src/hooks/amm.ts:
```diff
@@ -0,0 +1,95 @@
+ * @param deadline optional: the deadline for the txn to execute
+ * @returns amount of tokenTo received from the swap
+ */
+export async function swap(
```
should this execute a swap, or build a txn and return the txn?

aureliusbtc replied:
Let's separate out into two functions as discussed earlier today: 
-  One function that builds an unsigned / populated transaction e.g `return await swapContract.populateTransaction.swap`
- One function that utilizes the build tx to then execute signing on behalf of the caller, assuming a signing provider is attached

CryptoMaxPlanck committed:
fixup:
- remove unused param
- fix spacing

CryptoMaxPlanck commented on src/common/synapse_contracts.ts:
```diff
@@ -6,19 +6,29 @@ export namespace SynapseContracts {
     export class SynapseContract {
         readonly bridgeAddress:      string;
         readonly bridgeZapAddress?:  string;
+        readonly swapAddress?:       string;
```
does this work for the purpose of having a way to get the contract address of the swap contract? it can be manually added to each chain like the bridgeAddress and bridgeZapAddress' are

aureliusbtc replied:
lgtm

aureliusbtc commented on src/common/utils.ts:
```diff
@@ -90,4 +90,20 @@ const CHAINID_CONTRACTS_MAP: {[c: number]: SynapseContracts.SynapseContract} = {
+    )
+}
+
+export function uiToNative(amount: number, decimals: number): BigNumber {
```
Rename func to clarify what it does

aureliusbtc commented on src/common/utils.ts:
```diff
@@ -90,4 +90,20 @@ const CHAINID_CONTRACTS_MAP: {[c: number]: SynapseContracts.SynapseContract} = {
+    return BigNumber.from(Math.round(amount * Math.pow(10, decimals)));
+}
+
+export function nativeToUi(amount: number, decimals: number): number {
```
rename for clarity

aureliusbtc commented on src/hooks/index.ts:
```diff
@@ -1,3 +1,8 @@
+	swap,
+	calculateSwap,
+} from "./amm";
+
```
Concept of hooks is react specific - let's figure out where to rename the folder which'll hold this logic.

aureliusbtc commented on src/index.ts:
```diff
@@ -133,5 +135,7 @@ export {
-    TransactionError
+    TransactionError,
+    swap,
+    calculateSwap
```
We should create Types around the Swap contract here & the functions we expose to interact with them (so for now, calculateSwap & swap, also buildSwap when implemented(maybe with different naming))

aureliusbtc commented on src/hooks/amm.ts:
```diff
@@ -0,0 +1,90 @@
+ * @param amountFrom the amount of tokenFrom to swap
+ * @returns the amount that would be received from a swap of amountFrom
+ */
+export async function calculateSwap(
```
As discussed below - let's group these functions into a top-level Swap Type to create grouping of functions being exposed (we do this for Bridge right now here https://github.com/synapsecns/sdk/blob/master/src/bridge/bridge.ts#L57)

CryptoMaxPlanck committed:
fix:
- fix based on gh feedback
- add Swap contract interface
- use contract interface to get contract instance
- added provider pass for the txn creators
- add txn creators
- change return types
- fix imports and exports for index
- move amm code to designated folder

CryptoMaxPlanck committed:
add:
- fixup some formatting
- add swap addresses

aureliusbtc commented on src/amm/amm.ts:
```diff
@@ -0,0 +1,129 @@
+  deadline?: number
+): Promise<ContractTransaction> {
+  // Create the swap transaction
+  const swapTxn = await createSwapTxn(
```
This pattern looks good

CryptoMaxPlanck committed:
add:
- change format of amm.ts
- add addLiquidity function
- add get token by address function in tokens.ts

CryptoMaxPlanck commented on src/amm/amm.ts:
```diff
@@ -0,0 +1,206 @@
+  const swapContract = Swap__factory.connect(swapAddress, provider);
+
+  // Get the length of the token array and tokens associated with each index of `amounts`
+  const pooledTokensLength = 5;
```
need to find a good way to do this. look into if we can snag this directly onchain or if I need to loop through getToken(i) until an address is not returned.

CryptoMaxPlanck commented on src/amm/amm.ts:
```diff
@@ -0,0 +1,206 @@
+  // Build a populated transaction with the contract's `addLiquidity` function
+  return await swapContract.populateTransaction.addLiquidity(
+    amountsToNativeDecimals,
+    minToMint,
```
should this be converted to a certain specified decimals?

CryptoMaxPlanck committed:
add:
- removeLiquidity functions to amm.ts

CryptoMaxPlanck commented on src/amm/amm.ts:
```diff
@@ -0,0 +1,280 @@
+  const swapContract = Swap__factory.connect(swapAddress, provider);
+
+  // Get the length of the token array and tokens associated with each index of `minAmounts`
+  const pooledTokensLength = 5;
```
same as above.

CryptoMaxPlanck commented on src/amm/amm.ts:
```diff
@@ -0,0 +1,280 @@
+
+  // Build a populated transaction with the contract's `removeLiquidity` function
+  return await swapContract.populateTransaction.removeLiquidity(
+    amount,
```
same as above, need to convert units?

trajan0x commented:
Thanks for the contribution, but this repo is now deprecated and remaining issues being closed. The next iteration of the sdk can be found [here](https://github.com/synapsecns/sanguine/tree/master/packages/sdk-router)
